src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -52,36 +51,73 @@ public static ImmutableArray<BasicBlock> Create(IBlockOperation body)
             return blocks.ToImmutableAndFree();
         }
 
+        /// <summary>
+        /// Do a pass to eliminate blocks without statements that can be merged with predecessor(s).
+        /// </summary>
+        /// <param name="blocks"></param>
         private static void Pack(ArrayBuilder<BasicBlock> blocks)
         {
             int count = blocks.Count - 1;
             for (int i = 1; i < count; i++)
             {
                 BasicBlock block = blocks[i];
                 Debug.Assert(block.Next != null);
-                if (block.Statements.IsEmpty && block.Conditional.Value == null)
+                if (block.Statements.IsEmpty)
                 {
                     BasicBlock next = block.Next;
-                    foreach (BasicBlock predecessor in block.Predecessors)
+
+                    if (block.Conditional.Value == null)
```

Comment:
To help flatten the code, consider inverting this `if` with `if (... != null) { continue; }`